### PR TITLE
Add Available/Progressing conditions to DisaggregatedSet status

### LINKED ISSUE #868: [DisaggregatedSet] .status is never populated by the controller
/kind bug

**What happened**:

The `DisaggregatedSet` controller never writes `.status`. After create, rolling update, scale, and add/remove role, it stays empty:

```
$ kubectl get disaggregatedset ds-basic -o json | jq .status
{}
```

`observedGeneration` is also never set, so clients can't tell whether the controller has seen the latest `.spec`.

**What you expected to happen**:

Each reconcile should write at minimum:

- `status.observedGeneration`
- `status.replicas` / `readyReplicas` / `updatedReplicas`
- `status.conditions` (`Available` / `Progressing` / `Degraded`)

The CRD already declares the subresource and full schema in `api/disaggregatedset/v1/disaggregatedset_types.go` — this is a controller-side gap.

**How to reproduce it (as minimally and precisely as possible)**:

1. Apply any valid 2-role `DisaggregatedSet`.
2. Wait until pods are `Running`.
3. `kubectl get disaggregatedset <name> -o json | jq .status` → `{}`.

**Anything else we need to know?**:

Blocks `additionalPrinterColumns` on the CRD, and leaves "one role unschedulable" silently invisible to users.

**Environment**:

- Kubernetes version (`kubectl version`): v1.35.0 (kind v0.31.0, single-node)
- LWS version (`git describe --tags --dirty --always`): main @ commit `17ee230` (+ 2/N unified controller-manager merge)
- Cloud provider or hardware configuration: local kind cluster (bare-metal host)
- OS (`cat /etc/os-release`): Ubuntu 24.04.4 LTS (Noble Numbat)
- Kernel (`uname -a`): Linux 6.17.0-29-generic x86_64
- Install tools: kind + `make image-build` + `kubectl apply`
- Others: DisaggregatedSet CRD + controller co-located in `lws-controller-manager`

### PR DESCRIPTION
**Pull request by BenAyedMedAla**
/kind feature
/kind api-change

#### What this PR does / why we need it

Follow-up to #933, which added `status.observedGeneration` and per-role `status.roleStatuses`. Per @yankay's review on that PR, the `Available`/`Progressing` condition model was split out for its own focused design discussion:

- `Available`: every role has reached its desired replica count (its per-slice `spec.roles[].replicas`, or 1 by default, multiplied across all slices), with all of those replicas ready and updated to the current revision.
- `Progressing`: otherwise.

Two things I specifically addressed, since they were open questions in the original discussion:

- **Desired count, not just self-consistency.** A role whose running replicas are internally consistent (`ready == updated == running`) but below its desired count still reads as `Progressing`. Comparing against the desired count (rather than just checking the three counters agree with each other) is what's needed to get the next point right.
- **Zero-replica pause state.** `DisaggregatedSetSpec` has an `XValidation` allowing all roles to be paused at `replicas: 0`. With desired-count-based comparison, that state naturally reads as `Available` once fully drained (0 desired == 0 actual), rather than being stuck `Progressing` forever just because "desired" happens to be 0.

`Available` and `Progressing` are mutually exclusive. `LastTransitionTime` only changes on an actual `Status` flip (true → false or vice versa), not on every `ObservedGeneration` bump — this was flagged by Copilot on the original PR and is carried forward here with its regression tests.

#### Which issue(s) this PR fixes
Fixes #868

#### Special notes for your reviewer

- Deliberately no `Degraded` condition. Distinguishing "still rolling out" from "genuinely stuck/failed" needs a stronger signal (pod failure reasons, timeout heuristics) than what's derivable from the owned LWS statuses today. Happy to follow up if there's an existing convention in the codebase I should reuse instead of inventing one.
- Test coverage added for the two specific edge cases above (`TestStatusProgressingWhenUnderDesiredCount`, `TestStatusAvailableWhenPausedAtZero`), plus the existing fresh-deployment/fully-ready/role-removal cases extended with condition assertions.
- Regenerated CRD manifests and API reference docs to match (removes the "Degraded" bullet from the `Conditions` doc comment, since that condition type isn't implemented).

#### Does this PR introduce a user-facing change?
```release-note
DisaggregatedSet: the controller now reports an `Available`/`Progressing` condition in `status.conditions`, reflecting whether every role has reached its desired replica count with all replicas ready and updated to the current revision.
```

### CODE REVIEW

File: pkg/controllers/disaggregatedset/disaggregatedset_controller.go [hunk 1]
```diff
@@ -153,13 +154,16 @@ func (r *DisaggregatedSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
+// updateStatus recomputes per-role replica counts and the Available/Progressing
+// condition from the LWS objects the DisaggregatedSet owns (aggregated across all
+// slices and revisions), and persists the result if anything changed. roleNames is
+// always the current spec.roles: a role removed from spec has no RoleStatuses entry
+// even while its old LWS objects are still draining down to 0 (see RoleStatuses doc).
 func (r *DisaggregatedSetReconciler) updateStatus(ctx context.Context, disaggregatedSet *disaggregatedsetv1.DisaggregatedSet, roleNames []string, revision string) error {
 	roleStatuses := make([]disaggregatedsetv1.RoleStatus, 0, len(roleNames))
+	roleConfigs := disaggregatedsetutils.GetRoleConfigs(disaggregatedSet)
+	sliceCount := disaggregatedsetutils.GetSlices(disaggregatedSet)
+	available := true
```
[the remark this answers is not in the document]
Comment (BenAyedMedAla) [reply pkg/controllers/disaggregatedset/disaggregatedset_controller.go]: Good catch — fixed in 9fa971b. `updateStatus` now takes `scalers` and uses the same `getTargetReplicas` helper `reconcileRoleSimple` already relies on, so External roles compare against the scaler-resolved target instead of the ignored `spec.replicas`. Added `TestStatusUsesScalerTargetForExternalRoles` to pin it (creates an External role with a misleading inline `replicas: 5`, confirms the LWS is actually created at the scaler-seeded target of 1, and that `Available` correctly reflects that once ready — not stuck `Progressing` by comparing against the ignored 5).

File: pkg/controllers/disaggregatedset/disaggregatedset_controller.go [hunk 2]
```diff
@@ -203,6 +215,72 @@ func setRoleStatuses(disaggregatedSet *disaggregatedsetv1.DisaggregatedSet, role
...
+			}
+			continue
+		}
+		if cond.Status == metav1.ConditionTrue {
+			disaggregatedSet.Status.Conditions[i].Status = metav1.ConditionFalse
+			disaggregatedSet.Status.Conditions[i].LastTransitionTime = now
+			disaggregatedSet.Status.Conditions[i].ObservedGeneration = newCondition.ObservedGeneration
+			changed = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (BenAyedMedAla) [reply pkg/controllers/disaggregatedset/disaggregatedset_controller.go:309]: Good catch — fixed in 9fa971b. Both paths in `setDisaggregatedSetCondition` now sync `Reason`/`Message` along with `ObservedGeneration`, only leaving `LastTransitionTime` untouched when `Status` doesn't actually flip. Added `TestSetDisaggregatedSetCondition_SameStatusSyncsReasonAndMessage` and extended the existing flip test to assert on `Reason`/`Message` too.

[the remark this answers is not in the document]
Comment (BenAyedMedAla) [reply pkg/controllers/disaggregatedset/disaggregatedset_controller.go:309]: Follow-up on your suppressed comment about the flip clobbering any Status=True condition: fixed in 606ea6d. Added `exclusiveConditionTypes` (mirroring the existing pattern in `leaderworkerset_controller.go`) and scoped the flip to only the Available/Progressing pair, so a future condition type — or one written by another controller — can't get silently flipped just for being true. `TestSetDisaggregatedSetCondition_LeavesUnrelatedConditionTypeUntouched` pins this.

File: pkg/controllers/disaggregatedset/disaggregatedset_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenAyedMedAla) [reply pkg/controllers/disaggregatedset/disaggregatedset_controller.go]: Good follow-up catch — my comment was wrong and the fallback wasn't actually conservative. Fixed in 606ea6d: when an External role has no scaler in the map (e.g. a name collision with a foreign, non-owned `DisaggregatedSetRoleScaler` — same class of issue as #981, but for scalers), the role is now treated as explicitly `Progressing` rather than falling into `getTargetReplicas`'s literal-0 fallback, which could spuriously match a role that also has 0 actual replicas. `TestStatusProgressingWhenExternalRoleScalerMissing` reproduces the collision and pins the fix.

**Comment by ahg-g** [comment]
looks good, I will leave it to @yankay to approve 

/lgtm

**Comment by BenAyedMedAla** [comment]
Addressing Copilot's suppressed finding from the latest review (on `setDisaggregatedSetCondition`): the other condition in the exclusive pair was only ever touched by the Status-flip branch — once it settled to `False`, its `ObservedGeneration` froze there permanently even as generation kept advancing while the active condition stayed `True` with no real transition. That's misleading for a client reading `conditions[Progressing].observedGeneration` expecting it to track how current the reported state is.

Fixed in 65ef847: when the exclusive-pair condition is already `False` and stays `False`, only `ObservedGeneration` catches up (`Reason`/`Message` untouched, since they still accurately describe why it became false; `LastTransitionTime` untouched, since `Status` didn't transition). Added `TestSetDisaggregatedSetCondition_StaleFalseConditionObservedGenerationCatchesUp` to pin it.

**Comment by BenAyedMedAla** [comment]
CI failed across the board (unit/integration/verify/e2e) after the last push — root cause: #983 merged into `main` in the meantime, which changed `LeaderWorkerSetManager.Get`'s signature (`Get(ctx, namespace, name)` → `Get(ctx, *DisaggregatedSet, name)`). This branch hadn't been rebased since, so it was still compiling against the old signature once tested against current `main`.

Rebased onto `main` (170b898) — the three-way merge auto-resolved every pre-existing call site since #983 touched the same lines; only two calls added by this branch's own new test needed a manual fix. Full build, `go vet`, the whole test suite, and `golangci-lint` across the module all pass locally. Apologies that this reset the `lgtm` — happy to ping again once CI confirms green if that's useful.
